### PR TITLE
Pr mainloop testable

### PR DESCRIPTION
**Pull request by caleridas**
Rework mainloop to make it endpoint behavior decently testable.

After having done this, I am strongly inclined of ripping this sub-par event loop structure out completely and replacing it with something sensible ('asio', 'ace', 'libev', ... maybe 'libtscb' although I feel uncomfortable proposing my own code).

### CODE REVIEW

File: src/mavlink-router/mainloop.cpp [hunk 1]
```diff
@@ -37,69 +36,48 @@
...
+Mainloop* Mainloop::instance = nullptr;
 
-int Mainloop::open()
+Mainloop::Mainloop()
 {
-    if (epollfd != -1)
-        return -EBUSY;
+    if (instance) {
+        throw std::logic_error("Only one mainloop instance must exist at a given time");
+    }
```
Comment (jkflying) [inline src/mavlink-router/mainloop.cpp:47]: I guess somewhere you need a:

```suggestion
    }
    instance = this;
```

I didn't see it anywhere else

File: src/mavlink-router/mainloop.cpp [hunk 2]
```diff
@@ -373,6 +299,57 @@ void Mainloop::loop()
     }
 }
 
+int Mainloop::run_single(int timeout_msec)
+{
+    constexpr int max_events = 8;
+    struct epoll_event events[max_events];
```
Comment (jkflying) [inline src/mavlink-router/mainloop.cpp:307]: I guess this diff section is just a cut-paste-format across from the loop? No other changes?

File: src/mavlink-router/mainloop.cpp [hunk 3]
```diff
@@ -381,10 +358,8 @@ bool Mainloop::_log_aggregate_timeout(void *data)
         _errors_aggregate.msg_to_unknown = 0;
     }
 
-    if (g_endpoints) {
-        for (Endpoint **e = g_endpoints; *e != nullptr; e++) {
-            (*e)->log_aggregate(LOG_AGGREGATE_INTERVAL_SEC);
-        }
+    for (const auto& e : _endpoints) {
```
Comment (jkflying) [inline src/mavlink-router/mainloop.cpp:363]: Way way way cleaner, thank you!

File: src/mavlink-router/mainloop.h [hunk 4]
```diff
@@ -59,43 +79,47 @@ class Mainloop {
...
+    {
+        return _endpoints;
+    }
 
 private:
     static const unsigned int LOG_AGGREGATE_INTERVAL_SEC = 5;
 
     endpoint_entry *g_tcp_endpoints = nullptr;
-    Endpoint **g_endpoints = nullptr;
+    std::vector<std::unique_ptr<Endpoint>> _endpoints;
```
Comment (jkflying) [inline src/mavlink-router/mainloop.h:105]: Much nicer!

**Review by jkflying — APPROVED** [review]
Amazing cleanup. The only thing I saw which doesn't look right is the missing `instance = this;`

**Review by jkflying — DISMISSED** [review]

File: src/mavlink-router/mainloop.cpp (hunk 1 above)
Comment (caleridas) [reply src/mavlink-router/mainloop.cpp:47]: yes good catch that is wrong indeed -- I guess it should go to the very end, when all setup is done

File: src/mavlink-router/mainloop.cpp (hunk 2 above)
Comment (caleridas) [reply src/mavlink-router/mainloop.cpp:307]: the additional change is pulling out running the mainloop a bit longer for logger shutdown -- the way it was written originally was confusing as hell I think

**Comment by jkflying** [comment]
Ah I didn't notice, the tests are failing :-/

**Comment by caleridas** [comment]
> Ah I didn't notice, the tests are failing :-/

In my running it only fails for tests/routing-test.py, but that also failed before this change, I am not clear what the success conditions of this are (it is never run in any of our builds, so maybe it has been broken for long without anyone noticing/caring)

I will have a look what it is and maybe mark it is xfail if needed

**Comment by caleridas** [comment]
Made the python test skip on missing library instead of fail. This does not seem to be running anywhere presently -- arguably it would be nicer to check prerequisites in configure and not declare this a test case, but that is not so easily feasible.

**Review by jkflying — APPROVED** [review]

**Pull request merged**